File: Python/trie.py
class TrieNode:
    def __init__(self, children=None, is_end=False):
        self.children = children or {}
        self.is_end = is_end

    def __len__(self):
        return len(self.children)


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for ch in word:
            temp = current.children.get(ch, None)
            if not temp:
                temp = TrieNode()
                current.children[ch] = temp
            current = temp
        current.is_end = True

    def search(self, word):
        node = self.search_node(word)
        return node and node.is_end

    def starts_with(self, prefix):
        return self.search_node(prefix) is not None

    def search_node(self, word):
        current = self.root
        for ch in word:
            temp = current.children.get(ch, None)
            if not temp:
                return None
            current = temp
        return current

    def delete(self, word):
        self.delete_helper(self.root, word, 0)

    def delete_helper(self, current, word, index):
        if index == len(word):
            if not current.is_end:
                return False
            current.is_end = False
            return len(current.children) == 0

        ch = word[index]
        node = current.children.get(ch, None)
        if not node:
            return False
        should_delete_current_node = self.delete_helper(node, word, index + 1)
        if should_delete_current_node:
            del current.children[ch]
            return len(current.children) == 0
        return False


'''
# code below can be used to test 'Trie'

if __name__ == '__main__':
    trie = Trie()
    trie.insert('Fun')
    trie.insert('Food')
    trie.insert('Fame')
    trie.insert('Fault')
    print('Trie is', '' if trie.starts_with('F') else 'not', 'having words starting with F')
    print('Trie is', '' if trie.starts_with('D') else 'not', 'having words starting with D')
    print('Fun is', '' if trie.search('Fun') else 'not', 'present in trie.')
    print('Food is', '' if trie.search('Food') else 'not', 'present in trie.')
    print('Fame is', '' if trie.search('Fame') else 'not', 'present in trie.')
    print('Fault is', '' if trie.search('Fault') else 'not', 'present in trie.')
    trie.delete('Fault')
    print('Now "Fault" was removed from trie!')
    print('Fun is', '' if trie.search('Fun') else 'not', 'present in trie.')
    print('Food is', '' if trie.search('Food') else 'not', 'present in trie.')
    print('Fame is', '' if trie.search('Fame') else 'not', 'present in trie.')
    print('Fault is', '' if trie.search('Fault') else 'not', 'present in trie.')
'''

File: Java/Trie.java
import java.util.HashMap;

public class Trie {
    private class TrieNode {
        HashMap<Character, TrieNode> childern;
        boolean isEnd;

        TrieNode() {
            this.childern = new HashMap<>();
        }
    }

    private TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode current = this.root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            TrieNode temp = current.childern.get(ch);
            if (temp == null) {
                temp = new TrieNode();
                current.childern.put(ch, temp);
            }
            current = temp;
        }
        current.isEnd = true;
    }

    public boolean search(String word) {
        TrieNode trieNode = searchNode(word);
        return trieNode != null && trieNode.isEnd;
    }

    public boolean startsWith(String prefix) {
        return searchNode(prefix) != null;
    }

    private TrieNode searchNode(String word) {
        TrieNode current = this.root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);
            TrieNode temp = current.childern.get(ch);
            if (temp == null) return null;
            current = temp;
        }
        return current;
    }

    public void delete(String word) {
        delete(this.root, word, 0);
    }

    private boolean delete(TrieNode current, String word, int index) {
        if (index == word.length()) {
            if (!current.isEnd) return false;
            current.isEnd = false;
            return current.childern.size() == 0;
        }
        char ch = word.charAt(index);
        TrieNode node = current.childern.get(ch);
        if (node == null) return false;
        boolean shouldDeleteCurrentNode = delete(node, word, index + 1);
        if (shouldDeleteCurrentNode) {
            current.childern.remove(ch);
            return current.childern.size() == 0;
        }
        return false;
    }

    /* Commented code below can be used to test 'Trie'
    public static void main(String[] args) {
        Trie trie = new Trie();
        trie.insert("Fun");
        trie.insert("Food");
        trie.insert("Fame");
        trie.insert("Fault");
        System.out.println("Trie is " + (trie.startsWith("F") ? "" : "not ") + "having names starting with 'F'.");
        System.out.println("Trie is " + (trie.startsWith("D") ? "" : "not ") + "having names starting with 'D'.");
        System.out.println("Fun is " + (trie.search("Fun") ? "" : "not ") + "present in trie.");
        System.out.println("Food is " + (trie.search("Food") ? "" : "not ") + "present in trie.");
        System.out.println("Fame is " + (trie.search("Fame") ? "" : "not ") + "present in trie.");
        System.out.println("Fault is " + (trie.search("Fault") ? "" : "not ") + "present in trie.");
        trie.delete("Fault");
        System.out.println("Now 'Fault' is removed from trie.");
        System.out.println("Fun is " + (trie.search("Fun") ? "" : "not ") + "present in trie.");
        System.out.println("Food is " + (trie.search("Food") ? "" : "not ") + "present in trie.");
        System.out.println("Fame is " + (trie.search("Fame") ? "" : "not ") + "present in trie.");
        System.out.println("Fault is " + (trie.search("Fault") ? "" : "not ") + "present in trie.");
    }
    */
}
